### PR TITLE
Flaps/Spoiler refactor and enable for VTOL

### PR DESCRIPTION
**Pull request by sfuhrer**
Possible alternative to https://github.com/PX4/PX4-Autopilot/pull/19286

**Describe problem solved by this pull request**
Not able to use spoilers on VTOLs.

**Describe your solution**
commit by commit:
- spoiler support in hover LAND for VTOL, including slew rate limiting (1s) --> param `VT_SPOILER_MC_LD`
- enable automatic spoiler deploy in fixed-wing (based on in-descend flag or landing (which is not there atm, that would be some follow up work)) 
   - separate, fixed spoiler configuration for land and descend, `FW_SPOILERS_LND` and `FW_SPOILERS_DESC`
   - add trimming for spoiler deflection (`FW_DTRIM_P_SPOIL`)
   - 1s slew rate on spoiler output
   - remove "flaperon" output of FW attitude module, as it is actually not needed (if one want to use the ailerons as flaperons then one can just enable the flaps, and then in the mixer define that the aileron deflection is a sum of flaps+roll input)
 - add option for manual spoiler control in manual flight --> `FW_SPOILERS_MAN`
- move tilt to actuator_controls_1(6) (AIRBRAKES channel, was at the FLAPS channel before)
- use SlewRate function for flaps slew rate

**Describe possible alternatives**
tbd

**Test data / coverage**
Bench tested on a Tiltrotor VTOL with using the ailerons as spoilers.

**Additional context**
Missing:

- [x] figure out if we want to support flaperons, so ailerons deflecting down to increase lift (or if only deflecting them up for reducing lift is enough, so spoilers)
- [x] rename "flaperon" to "spoiler" if the above is ok
- [x] add differential trimming based on spoiler setpoint (similar to what we have for flaps atm)
- [x] VTOL: add option to set flap and spoiler setpoint in back transition
- [x] VTOL: add option to set spoiler setpoint in hover (guess flaps are less useful?)
- [ ] FW: add logic around using spoilers do increase descent rate (reducing lift --> increases flight speed and with it descent rate)

### CODE REVIEW

**Comment by ryanjAA** [comment]
I was going through this today and tested it out on the bench with a pixhawk and servos attached (not sitl but on hardware and during front transition the flaps deploy (not sure how to turn that off) and on back back transition the ailerons setup up in the mixer as airbrake (spoilers) don't deploy. When in FW flight if I initiate the flaps button on the RC remote flaps deploy and spoilers (ailerons) do go up.

**Comment by Jaeyoung-Lim** [comment]
@sfuhrer Is there a way we can get a configuration in simulation to test this? Since the `plane` model has flaps would this be sufficient? (maybe crowflaps for spoilers?)

**Comment by ryanjAA** [comment]
@Jaeyoung-Lim i was thinking the same thing. The current SITL VTOL is rather limited so it would be good to test more functionality in the sim.

**Comment by sfuhrer** [comment]
BTW, it is not yet enabled with the new control allocation, I'll look into that next.

**Comment by sfuhrer** [comment]
I've started testing it on a Tiltrotor VTOL. Basically now deflect the ailerons in hover land and fixed-wing descent (fixed-wing descend logic is still missing here atm). So far the logic seems to function correctly.

@dagar @RomanBapst could I get your initial feedback here? Most of all, would you say it's worth to move the combined tilt from `actuator_controls_1[4]` to `actuator_controls_1[6] `for the sake of enabling flaps and spoilers on Tiltrotor VTOLs? That will break existing mixer files - though I guess as now we anyway plan to get people to the new control allocation system the timing probably isn't too bad (best after the next release is branched off, if it isn't already).
Further: 
- remove the flaperon on `actuator_controls_1[5]` --> this will make it impossible to use flaps + flaperons that can be deflected both up and down, though it's still totally possible to map the ailerons to **either** the flaps **or** the spoiler channel in Auto flight. In manual flight the functionality is preserved anyway.

**Comment by ryanjAA** [comment]
>Further:
> 
> * remove the flaperon on `actuator_controls_1[5]` --> this will make it impossible to use flaps + flaperons that can be deflected both up and down, though it's still totally possible to map the ailerons to **either** the flaps **or** the spoiler channel in Auto flight. In manual flight the functionality is preserved anyway, they can be mapped to a 3-way switch that allows to deflect them both up and down.


So am I correct in understanding that I cant have my flaps down and ailerons up now during landing or back transition? Both in auto and manual mode?

File: src/modules/vtol_att_control/standard.cpp [hunk 1]
```diff
@@ -370,8 +378,9 @@ void Standard::fill_actuator_outputs()
 		fw_out[actuator_controls_s::INDEX_PITCH]        = elevon_lock ? 0 : fw_in[actuator_controls_s::INDEX_PITCH];
 		fw_out[actuator_controls_s::INDEX_YAW]          = 0;
 		fw_out[actuator_controls_s::INDEX_THROTTLE]     = _pusher_throttle;
-		fw_out[actuator_controls_s::INDEX_FLAPS]        = 0;
-		fw_out[actuator_controls_s::INDEX_AIRBRAKES]    = 0;
+		fw_out[actuator_controls_s::INDEX_FLAPS]        = _flaps_setpoint_with_slewrate.getState();
```
Comment (RomanBapst) [inline src/modules/vtol_att_control/standard.cpp:381]: @sfuhrer Where are you resetting these for hover? Do I understand correctly that if you have flaps/spoilers set for backtransition then they will be kept in hover mode?

File: src/modules/vtol_att_control/standard.cpp [hunk 2]
```diff
@@ -409,7 +419,8 @@ void Standard::fill_actuator_outputs()
 		fw_out[actuator_controls_s::INDEX_PITCH]        = fw_in[actuator_controls_s::INDEX_PITCH];
 		fw_out[actuator_controls_s::INDEX_YAW]          = fw_in[actuator_controls_s::INDEX_YAW];
 		fw_out[actuator_controls_s::INDEX_THROTTLE]     = fw_in[actuator_controls_s::INDEX_THROTTLE];
-		fw_out[actuator_controls_s::INDEX_FLAPS]        = fw_in[actuator_controls_s::INDEX_FLAPS];
+		fw_out[actuator_controls_s::INDEX_FLAPS]        = _flaps_setpoint_with_slewrate.getState();
```
Comment (RomanBapst) [inline src/modules/vtol_att_control/standard.cpp:422]: @sfuhrer In FW mode shouldn't you just feedforward the actuator controls from the FW module?

Comment (sfuhrer) [reply src/modules/vtol_att_control/standard.cpp:422]: But then we'd possibly have a step in the controls when the vehicle switches from a transition to FW and the other way around. What do we lose by always going through the slew rate in the VTOL module like proposed here?

File: src/modules/vtol_att_control/standard.cpp (hunk 1 above)
Comment (sfuhrer) [reply src/modules/vtol_att_control/standard.cpp:381]: In `void VtolType::update_mc_state()`
`_flaps_setpoint_with_slewrate.update(0.f, _dt);`

Comment (ryanjAA) [reply src/modules/vtol_att_control/standard.cpp:381]: Side note on this part - don't we want the least amount of drag in hover (so if wind is there it easier to manage)? Meaning doesn't it make more sense to not have anything deflected while in hover?

**Comment by ryanjAA** [comment]
> > So am I correct in understanding that I cant have my flaps down and ailerons up now during landing or back transition? Both in auto and manual mode?
> 
> You totally can have that. But it wouldn't be possible to e.g. deflect the ailerons down for takeoff and up for landing, they're either linked to the flaps or to the spoiler channel. Without code changes that though also wasn't possible till now, so I don't think it's something that's used anyway.

Ok - good to know; that is what is expected then. While full wing flaps would be great for taking off in traditional FW for a very heavy setup that hasnt ever been possible so we are all synced up then on that. Thanks!

File: src/modules/vtol_att_control/standard.cpp (hunk 1 above)
Comment (sfuhrer) [reply src/modules/vtol_att_control/standard.cpp:381]: Yes, but we also a don't want a lot of lift when descending, and the spoilers can help with that. The small increase in drag doesn't hurt much I would say.
And keep in mind that with what I propose the spoilers are only deflected when landing, not just in general.

**Comment by ryanjAA** [comment]
Maybe I'm missing something but I have this setup and still don't have the ailerons deflecting up when I manually hit the flaps switch. See attached mixer and params. Ailerons are mixed to control output 5 (spoilers).

Also, can the slew rate for flaps be altered? the ~1-2 seconds isn't a problem - slow is good but it takes about 1 second for them to start moving.
[Vtol_test_params.txt](https://github.com/PX4/PX4-Autopilot/files/8421764/Vtol_test_params.txt)
[vtol_test.main.mix.txt](https://github.com/PX4/PX4-Autopilot/files/8421766/vtol_test.main.mix.txt)

File: src/modules/fw_att_control/FixedwingAttitudeControl.cpp [hunk 3]
```diff
@@ -67,6 +67,8 @@ FixedwingAttitudeControl::FixedwingAttitudeControl(bool vtol) :
 	_yaw_ctrl.set_max_rate(radians(_param_fw_acro_z_max.get()));
 
 	_rate_ctrl_status_pub.advertise();
+	_spoiler_setpoint_with_slewrate.setSlewRate(1.f); //min 1s from no spoiler deflection to full deflection
+	_flaps_setpoint_with_slewrate.setSlewRate(1.f);
```
Comment (tstastny) [inline src/modules/fw_att_control/FixedwingAttitudeControl.cpp:71]: can we make these "1.f" named constexpr?

File: src/modules/fw_att_control/FixedwingAttitudeControl.cpp [hunk 4]
```diff
@@ -726,40 +734,44 @@ void FixedwingAttitudeControl::control_flaps(const float dt)
...
-	float flaperon_control = 0.0f;
+	if (_vcontrol_mode.flag_control_manual_enabled) {
+		switch (_param_fw_spoilers_man.get()) {
+		case 0:
+			break;
 
-	/* map flaperons by default to manual if valid */
-	if (PX4_ISFINITE(_manual_control_setpoint.aux2) && _vcontrol_mode.flag_control_manual_enabled
-	    && fabsf(_param_fw_flaperon_scl.get()) > 0.01f) {
+		case 1:
```
Comment (tstastny) [inline src/modules/fw_att_control/FixedwingAttitudeControl.cpp:749]: after a quick look, im guessing the answer to this is no, so this is probably a discussion for another issue/PR - but is there currently functionality in the param system to get any enumeration? (e.g. set from the `@value`) - would be nice to pull the enums directly from the param definition and use them in the code. e.g. something like `case param_fw_spoilers_man_::kFlapsChannel:`

File: src/modules/fw_att_control/FixedwingAttitudeControl.hpp [hunk 5]
```diff
@@ -231,6 +235,7 @@ class FixedwingAttitudeControl final : public ModuleBase<FixedwingAttitudeContro
 	ECL_WheelController		_wheel_ctrl;
 
 	void control_flaps(const float dt);
+	void control_spoilers(const float dt);
```
Comment (tstastny) [inline src/modules/fw_att_control/FixedwingAttitudeControl.hpp:238]: I know I'm being annoying - but while we're adding new methods, can we incrementally start adding the `@brief` and `@param[in]` description + units here? `dt` could mean many things to someone new to the code.

File: src/modules/fw_att_control/fw_att_control_params.c [hunk 6]
```diff
@@ -472,7 +474,8 @@ PARAM_DEFINE_FLOAT(FW_FLAPS_SCL, 1.0f);
 /**
  * Flaps setting during take-off
  *
- * Sets a fraction of full flaps (FW_FLAPS_SCL) during take-off
+ * Sets a fraction of full flaps (FW_FLAPS_SCL) during take-off.
```
Comment (tstastny) [inline src/modules/fw_att_control/fw_att_control_params.c:477]: oo thanks for adding the clarification on this param.. i actually misunderstood it in the last weeks! thought it was percentage of full. not a fraction of the already scaled by `FW_FLAPS_SCL` param... 

which actually leads me to another thought.. why? should we just go ahead and change this now to mean fraction of 100%? wouldn't this be much easier to understand? -- same for landing scale. Or maybe I'm missing a relevant justification for the "scale of the scale" approach.

File: src/modules/fw_att_control/fw_att_control_params.c [hunk 7]
```diff
@@ -735,3 +727,57 @@ PARAM_DEFINE_FLOAT(FW_DTRIM_R_FLPS, 0.0f);
...
+
+/**
+ * Spoiler input in manual flight
+ *
+ * Chose source for manual setting of spoilers in manual flight modes.
+ * Also applies to spoilerons (ailerons deflected down) if enabled in the mixer/allocation.
+ *
+ * @value 0 Disabled
+ * @value 0 Flaps channel
+ * @value 0 Aux1
```
Comment (tstastny) [inline src/modules/fw_att_control/fw_att_control_params.c:780]: these are all 0 -- supposed to be 0,1,2, right?

File: src/modules/vtol_att_control/vtol_type.cpp [hunk 8]
```diff
@@ -136,6 +136,9 @@ bool VtolType::init()
 		}
 	}
 
+	_spoiler_setpoint_with_slewrate.setSlewRate(1.f); //at least 1s from no spoiler deflection to full deflection
+	_flaps_setpoint_with_slewrate.setSlewRate(1.f); //at least 1s from no flaps deflection to full deflection
+
```
Comment (tstastny) [inline src/modules/vtol_att_control/vtol_type.cpp:141]: same comment - can we make `1.f` named constexpr?

File: src/modules/vtol_att_control/vtol_type.h [hunk 9]
```diff
@@ -207,6 +211,9 @@ class VtolType
 
 	virtual void parameters_update() = 0;
 
+	void setDt(float dt) {_dt = dt; }
```
Comment (tstastny) [inline src/modules/vtol_att_control/vtol_type.h:214]: documentation with brief, `param[in]` and units pls.

**Review by tstastny — CHANGES_REQUESTED** [review]
mostly minor comments. thanks for also cleaning things up a bit as you go (e.g. adding the enums for actuator controls) ;) fw_pos_ctrl needs all the incremental help it can get.

File: src/modules/fw_att_control/FixedwingAttitudeControl.cpp (hunk 3 above)
Comment (sfuhrer) [reply src/modules/fw_att_control/FixedwingAttitudeControl.cpp]: https://github.com/PX4/PX4-Autopilot/pull/19329/commits/98ea2aeb1c80f716c5335c194fbe02116df85ca6

File: src/modules/fw_att_control/FixedwingAttitudeControl.hpp (hunk 5 above)
Comment (sfuhrer) [reply src/modules/fw_att_control/FixedwingAttitudeControl.hpp]: https://github.com/PX4/PX4-Autopilot/pull/19329/commits/98ea2aeb1c80f716c5335c194fbe02116df85ca6

File: src/modules/fw_att_control/fw_att_control_params.c (hunk 7 above)
Comment (sfuhrer) [reply src/modules/fw_att_control/fw_att_control_params.c]: https://github.com/PX4/PX4-Autopilot/pull/19329/commits/98ea2aeb1c80f716c5335c194fbe02116df85ca6

File: src/modules/fw_att_control/FixedwingAttitudeControl.cpp (hunk 4 above)
Comment (sfuhrer) [reply src/modules/fw_att_control/FixedwingAttitudeControl.cpp:749]: I don't think it's possible atm, but would be cool to have. @bkueng am I right that it doesn't exist atm?

File: src/modules/vtol_att_control/vtol_type.cpp (hunk 8 above)
Comment (sfuhrer) [reply src/modules/vtol_att_control/vtol_type.cpp]: https://github.com/PX4/PX4-Autopilot/pull/19329/commits/98ea2aeb1c80f716c5335c194fbe02116df85ca6

File: src/modules/vtol_att_control/vtol_type.h (hunk 9 above)
Comment (sfuhrer) [reply src/modules/vtol_att_control/vtol_type.h:222]: https://github.com/PX4/PX4-Autopilot/pull/19329/commits/98ea2aeb1c80f716c5335c194fbe02116df85ca6

File: src/modules/fw_att_control/fw_att_control_params.c (hunk 6 above)
Comment (sfuhrer) [reply src/modules/fw_att_control/fw_att_control_params.c:477]: I didn't want to change everything at once:) Here it really should be about enabling spoiler and flaps for VTOLs, let's follow up with another PR for more refactoring. 
I though 100% agree with you though, I would remove `FW_FLAPS_SCL` completely.

Comment (sfuhrer) [reply src/modules/fw_att_control/fw_att_control_params.c]: Changed my mind, here it is: https://github.com/PX4/PX4-Autopilot/pull/19329/commits/3689864c2a2a6c8bb0fd2edede65ebc74fe61c1a

**Comment by sfuhrer** [comment]
@ryanjAA sorry for the delayed answer, I totally missed your comment before.

>Maybe I'm missing something but I have this setup and still don't have the ailerons deflecting up when I manually hit the flaps switch. See attached mixer and params. Ailerons are mixed to control output 5 (spoilers).

I you want the ailerons to move on Flaps commands then you'd need to map them to the flap channel, so for the aileron mixer:
M: 2
S: 1 0  10000  10000      0 -10000  10000
S: 1 4  10000  10000      0 -10000  10000

>Also, can the slew rate for flaps be altered? the ~1-2 seconds isn't a problem - slow is good but it takes about 1 second for them to start moving.
It should start immediately and be done after 1second currently, when I tested it on a VTOL this seemed to be correct. I should probably test it on a fixed-wing as well.

**Comment by ryanjAA** [comment]
Thanks @sfuhrer ill try again both way and post results. There is definitely a delay though from when hitting the flap button I’ll see if there is a difference in RC input to output outside of what you just explained.

File: src/modules/fw_att_control/FixedwingAttitudeControl.hpp [hunk 10]
```diff
@@ -230,7 +236,19 @@ class FixedwingAttitudeControl final : public ModuleBase<FixedwingAttitudeContro
...
+	 * @param dt Current time delta [s]
+	 */
+	void controlFlaps(const float dt);
+
+	/**
+	 * @brief Set current time delta for spoiler control (used in slew rate limiter)
+	 *
+	 * @param dt Current time delta [s]
+	 */
+	void controlSpoilers(const float dt);
```
Comment (tstastny) [inline src/modules/fw_att_control/FixedwingAttitudeControl.hpp:251]: Could we make the brief say something like "update flap (spoiler) control setting"? as it is not only setting the time delta, but also changing the commanded flap / spoiler value.

File: src/modules/fw_att_control/FixedwingAttitudeControl.hpp [hunk 11]
```diff
@@ -78,6 +79,9 @@ using uORB::SubscriptionData;
 
 using namespace time_literals;
 
+static constexpr float kFlapSlewRate = 1.f; //minimum 1 second from nono to full flap deflection
+static constexpr float kSpoilerSlewRate = 1.f; //minimum 1 second from nono to full spoiler deflection
```
Comment (tstastny) [inline src/modules/fw_att_control/FixedwingAttitudeControl.hpp:83]:
```suggestion
static constexpr float kFlapSlewRate = 1.f; //minimum 1 second from none to full flap deflection
static constexpr float kSpoilerSlewRate = 1.f; //minimum 1 second from none to full spoiler deflection
```

File: src/modules/vtol_att_control/vtol_att_control_main.cpp [hunk 12]
```diff
@@ -401,6 +409,7 @@ VtolAttitudeControl::Run()
 
 #endif // !ENABLE_LOCKSTEP_SCHEDULER
 
+	const float dt = math::min((now - _last_run_timestamp) / 1e6f, 0.1f);
```
Comment (tstastny) [inline src/modules/vtol_att_control/vtol_att_control_main.cpp:412]: pls define `0.1f` -- e.g. `kMaxVTOLAttitudeControlTimeStep`

File: src/modules/vtol_att_control/vtol_type.h [hunk 13]
```diff
@@ -292,6 +307,18 @@ class VtolType
...
+	 * @brief      Overrides controls for control surface or tilt mechanism in the context of preflight tests.
+	 *
+	 * @return     true while executing the test, false once done
+	 */
+	bool override_controls_for_test_mode();
+
+	SlewRate<float> _spoiler_setpoint_with_slewrate;
+	SlewRate<float> _flaps_setpoint_with_slewrate;
+
+	float _dt{0.0025f};
```
Comment (tstastny) [inline src/modules/vtol_att_control/vtol_type.h:320]: units and description pls :)

**Review by tstastny — CHANGES_REQUESTED** [review]
just a couple more small things i spotted

File: src/modules/fw_att_control/FixedwingAttitudeControl.hpp (hunk 10 above)
Comment (sfuhrer) [reply src/modules/fw_att_control/FixedwingAttitudeControl.hpp:251]: yes ofc, that was a stupid copy&paste thingy.

Comment (sfuhrer) [reply src/modules/fw_att_control/FixedwingAttitudeControl.hpp]: https://github.com/PX4/PX4-Autopilot/pull/19329/commits/3dd4403463f210743ac80a5d838a7de88f934259

File: src/modules/fw_att_control/FixedwingAttitudeControl.hpp (hunk 11 above)
Comment (sfuhrer) [reply src/modules/fw_att_control/FixedwingAttitudeControl.hpp]: https://github.com/PX4/PX4-Autopilot/pull/19329/commits/3dd4403463f210743ac80a5d838a7de88f934259

File: src/modules/vtol_att_control/vtol_att_control_main.cpp (hunk 12 above)
Comment (sfuhrer) [reply src/modules/vtol_att_control/vtol_att_control_main.cpp]: https://github.com/PX4/PX4-Autopilot/pull/19329/commits/3dd4403463f210743ac80a5d838a7de88f934259

File: src/modules/vtol_att_control/vtol_type.h (hunk 13 above)
Comment (sfuhrer) [reply src/modules/vtol_att_control/vtol_type.h]: https://github.com/PX4/PX4-Autopilot/pull/19329/commits/3dd4403463f210743ac80a5d838a7de88f934259

**Review by tstastny — DISMISSED** [review]

**Comment by sfuhrer** [comment]
I've cleaned up the commits and additionally moved the combined tilt to controls[8] after syncing with @dagar on that. To have it on a dedicated controls index is for sure cleaner than what we had till now (using [4], so the flaps index).
Additionally I think it makes most sense to only allow positive controls for flaps, spoilers and airbrakes, as they can only deploy in one direction. I've addressed that by constraining the controls_flaps/spoilers (update into slew rate) to [0, 1], see https://github.com/PX4/PX4-Autopilot/pull/19329/commits/45d87f7b699fa5049835d172725f8985a68b97a9.
I've removed from this PR some VTOL functionality that isn't yet clear if needed and opened a new PR for it: https://github.com/PX4/PX4-Autopilot/pull/19550
@tstastny please have a look at the last two comments, the remaining should only have changed marginaly since you last reviewed.

**Comment by tstastny** [comment]
> @tstastny please have a look at the last two comments, the remaining should only have changed marginaly since you last reviewed

lgtm

**Review by tstastny — DISMISSED** [review]

**Comment by sfuhrer** [comment]
Force-pushed to fix small format error.

**Review by RomanBapst — APPROVED** [review]

**Pull request merged**